Sequence.cpp

```diff
@@ -2,96 +2,136 @@
 
 void Sequence::begin_sequence() {
+    
+    int T0 = 0;                         
+    int T1 = T0 + DELAY_DROP_TONE_MS;   
+    int T2 = T1 + DELAY_INTERVAL_MS;     
+    int T3 = T2 + DELAY_DROP_TONE_MS;  
+    int T4 = T3 + DELAY_INTERVAL_MS;    
+    int T5 = T4 + DELAY_DROP_TONE_MS;  
+    int T6 = T5 + DELAY_INTERVAL_MS;   
+    int T7 = T6 + GATE_ACTIVE_MS;       
+    int T8 = T6 + DELAY_DROP_TONE_FINAL_MS; 
+
+    int target[] = {T0, T1, T2, T3, T4, T5, T6, T7, T8};
 
-  // gate timings
-  // Tone Play Time, wait length, light tree step
-  gate_step gate_steps[] =  {
-    { DELAY_DROP_TONE_MS, 0, 1 },
-    { 0, 60, 0 },
-    { DELAY_DROP_TONE_MS, 0, 2 },
-    { 0, 60, 0 },
-    { DELAY_DROP_TONE_MS, 0, 3 },
-    { 0, 60, 0 },
-    { DELAY_DROP_TONE_FINAL_MS, 0, 4 },
-    { 0, 0, 0 }  
-  };
+    int seq[][3] = { 
+      { 1, 1,-1},
+      {-1, 0,-1},
+      { 2, 1,-1}, 
+      {-1, 0,-1},
+      { 3 ,1,-1},
+      {-1, 0,-1},
+      { 4, 1, 1},
+      {-1,-1, 0},
+      {-1, 0,-1},
+    };
+
+    int count = 9;
+    int step = 0;
+    unsigned long offset;
+    unsigned long now = 0;
+    int val;
 
   serial_print("Sequence begin");
 
   gate->set_sequence_running(true);
-  int step = 0;
-  int playing_tone = 0;
-  int waiting = 0;
-  unsigned long timer_start = 0;
-  int gate_step_count = 7; // can't programatically count dynamic array
 
-  digitalWrite(LED_BUILTIN, HIGH);  // turn on LED
-  digitalWrite(PIN_LED_ACTIVE, HIGH);  // turn on LED
+  digitalWrite(LED_BUILTIN, HIGH);  // turn on builtin LED
+  digitalWrite(PIN_LED_ACTIVE, HIGH);  // turn on "Active" LED
 
   lighttree->led_reset();
   audio->play_sound_samples();
   gate->random_wait();
 
+  // once the sequence starts, do not allow abort
   serial_print("No longer abortable");
-  gate->set_abortable(false); // once the sequence starts, do not allow abort
-  while((step < gate_step_count) && (!gate->is_aborted())) {
-    unsigned long now = millis();
-    unsigned long timer = now - timer_start;
-
-    if (playing_tone && (timer >= gate_steps[step].tone_length)) {
-      // reset timer
-      playing_tone = 0;
-      serial_print_val("Stop tone",step);
-      serial_print_val("Tone length",gate_steps[step].tone_length);
-      audio->stop_tone();
-      step++;
-    }
-    if (!playing_tone && (gate_steps[step].tone_length > 0)) {
-      timer_start = now;
-      audio->start_tone(TONE_DROP_HZ);
-      serial_print_val("LED step",step);
-      lighttree->light_set(gate_steps[step].light_num, gate);
-      playing_tone = 1;
-    }
-    if (waiting && (timer >= gate_steps[step].wait_length)) {
-      waiting = 0;
-      step++;
-    }
-    if (!waiting && (gate_steps[step].wait_length > 0)) {
-      timer_start = now;
-      waiting = 1;
+  gate->set_abortable(false); 
+
+  // Use interrupt to catch END request from SENSOR
+  attachInterrupt(digitalPinToInterrupt(PIN_SENSOR), Interrupt2, LOW);
+  
+  offset = millis();
+  
+  while((step < count) && (!gate->is_aborted())) {
```

Comment:
@lawabb what was the reason for the refactor with the loop? was the original code difficult / unwieldily to update? (if the original code was rubbish, please say so!)

Reply:
The original code was fine. It took me sometime to figure out how it worked, but I successfully modified it to work with my addition of 'gate deactivate' with the exception that I couldn't set wait time to zero so had to use 1ms. The trouble was that it was becoming unweildy and ugly... I don't know how to merge with the original code.  As an amateur python programmer what I did was write the sequence in python the transferred it virtually line by line with help from google.  It may be of some use...
[test_new_seq.zip](https://github.com/dfrencham/rad-gate/files/1783351/test_new_seq.zip)